File: src/TorrentInfo.cpp
#include "TorrentInfo.h"

torrent_info_ptr torrent_info_create(const char *filename, int flags) {
    torrent_info_ptr ti = new torrent_info(string(filename), flags);

    return ti;
}

void torrent_info_release(torrent_info_ptr ti) {
    delete ti;
}

long total_size(torrent_info_ptr ti) {
    return ti->total_size();
}

int piece_length(torrent_info_ptr ti) {
    return ti->piece_length();
}

int num_pieces(torrent_info_ptr ti) {
    return ti->num_pieces();
}

const char *info_hash(torrent_info_ptr ti) {
    return ti->info_hash().to_string().c_str();
}

#ifdef JNI_INTERFACE_ENABLED

#define JNI_METHOD_BEGIN_TI(type, name, ...) \
    JNI_METHOD_BEGIN_HANDLE(type, name, torrent_info_ptr, ti, ##__VA_ARGS__)

JNI_METHOD_BEGIN(jlong, create, jstring filename, int flags)

        jboolean isStrCopy;
        const char *cFilename = env->GetStringUTFChars(filename, &isStrCopy);

        torrent_info_ptr ti = torrent_info_create(cFilename, flags);

        env->ReleaseStringUTFChars(filename, cFilename);

        return (jlong) ti;

JNI_METHOD_END_RET(jlong)

JNI_METHOD_BEGIN_TI(void, release)

        torrent_info_release(ti);

JNI_METHOD_END

JNI_METHOD_BEGIN_TI(jlong, totalSize)

        return total_size(ti);

JNI_METHOD_END_RET(jlong)

JNI_METHOD_BEGIN_TI(jint, pieceLength)

        return piece_length(ti);

JNI_METHOD_END_RET(jint)

JNI_METHOD_BEGIN_TI(jint, numPieces)

        return num_pieces(ti);

JNI_METHOD_END_RET(jint)

JNI_METHOD_BEGIN_TI(jstring, infoHash)

        return env->NewStringUTF(info_hash(ti));

JNI_METHOD_END_RET(jstring)

#endif //JNI_INTERFACE_ENABLED

File: src/com/frostwire/libtorrent/TorrentInfo.java
package com.frostwire.libtorrent;

import com.frostwire.libtorrent.swig.libtorrent;
import com.frostwire.libtorrent.swig.torrent_info;

public class TorrentInfo {

    private final torrent_info inf;

    public TorrentInfo(String filename) {
        this.inf = new torrent_info(filename);
    }

    public long totalSize() {
        return this.inf.total_size();
    }

    public int pieceLength() {
        return this.inf.piece_length();
    }

    public int numPieces() {
        return this.inf.num_pieces();
    }

    public String infoHash() {
        return libtorrent.to_hex(this.inf.info_hash().to_string());
    }

    public torrent_info getInf() {
        return this.inf;
    }

    public String mkString() {
        StringBuilder sb = new StringBuilder();

        sb.append("total_size = " + totalSize() + System.lineSeparator());
        sb.append("piece_length = " + pieceLength() + System.lineSeparator());
        sb.append("num_pieces = " + numPieces() + System.lineSeparator());
        sb.append("info_hash = " + infoHash() + System.lineSeparator());

        return sb.toString();
    }
}
